resolve.c

```diff
@@ -119,17 +139,36 @@ void resolveForwardDestinations(bool onlynew)
 		// Lock data when obtaining IP of this forward destination
 		lock_shm();
 		const char* ipaddr = getstr(forwarded[forwardID].ippos);
+		const char* name = getstr(forwarded[forwardID].namepos);
 		unlock_shm();
 
 
 		// Important: Don't hold a lock while resolving as the main thread
 		// (dnsmasq) needs to be operable during the call to resolveHostname()
-		const char* hostname = resolveHostname(ipaddr);
+		char* hostname = resolveHostname(ipaddr);
 
 		// Finally, lock data when storing obtained hostname
 		lock_shm();
-		forwarded[forwardID].namepos = addstr(hostname);
+		// Only store new hostname if it changed
+		if(hostname != NULL && strcmp(name, hostname) != 0)
+		{
+			// We do not need to check for name == NULL as name is
+			// always initialized with an empty string at position 0
+			forwarded[forwardID].namepos = addstr(hostname);
+		}
+		else if(config.debug & DEBUG_SHMEM)
+		{
+			// Debugging output
+			logg("Not adding \"%s\" to buffer (unchanged)", name);
+		}
```

Comment:
This code is almost duplicated, maybe it is possible to extract it into a function.

Reply:
Is this comment meant to only spawn the `logg()` routine or dis you mention this with respect to the whole functions? If it is the latter, then yes, I agree. However, I don't think it is enough code to be worth putting into a common function that would then also need some pointer magic to be able to fill the `namepos` fields of once the `clients` and once the `forwarded` data structures.

Reply:
The code from line 139 to 172 (excluding line 169) can be easily extracted into a function with the following signature (the name is up for debate):
```c
size_t resolveAndAddHostname(size_t ippos, size_t namepos);
```

And would be used in the following way (example given is for forward destinations):
```c
forwarded[forwardID].namepos = resolveAndAddHostname(
    forwarded[forwardID].ippos,
    forwarded[forwardID].namepos
);

forwarded[forwardID].new = false;
```

If the hostname has changed then the output of the function would be the output of `addstr`. If it has not changed, then the output would be the original `namepos`.